src/util/rateLimiters.tsx

```diff
@@ -0,0 +1,66 @@
+import { useCallback, useMemo, useRef } from 'react'
```

Comment:
This file defines two hooks for closely related rate limiters which we've wound up using a lot.
A *debouncer* collects a user input stream and holds off on processing it until it has been quiet for a particular period of time (e.g. type-ahead search not rerunning the search until the user has stopped typing for 200 ms).
A *throttler* keeps a function from updating more than once every x (while still changing the ultimate effect of the update up until the time it's run).

These are essentially the same rate-limiting function, the question is just do we cancel and reschedule while input is incoming (debouncer) or do we leave the scheduled update, but not schedule a new one until that one has cleared (throttler).

Anyway, I've decomposed both of these types into a couple pieces. From my perspective, both consist of an *updater* function (which tracks the new target state and handles scheduling of its resolution) and a *resolver* function (which actually updates the permanent state), as well as a set of *refs* (which can be modified rapidly without triggering rerenders or updates).

The general flow is that the refs are created and persisted, the updater changes the values in the refs and schedules resolution if the update makes the state different from the persistent state, and the resolver (when it finally gets to run) makes the persisted state look like what's in the refs.

Probably a good practical example is in the `views/common/Animation/AnimationControls/useWheelHandler.tsx` file, which handles throttled updates to the current playback frame as the user scrolls through with the mouse wheel. We only update the actual current frame a maximum of 60 times per second, no matter how fast you move the scroll wheel; but we do collect a count of the number of scrolls in any particular direction in order to determine how much to adjust the frame state.